### DISCUSSION TITLE
bind: cannot assign requested address

### DISCUSSION THREAD

**Discussion by Taymindis** [lima-vm/lima#1080] [Q&A] [answered]
### Description

192.168.1.105 is my local private ip, but it's not able to assigned

```bash
lima nerdctl run -d --name nginx -p 192.168.1.105:8809:80 nginx:alpine
FATA[0000] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error 
during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2022-09-
30T09:47:43Z" level=fatal msg="listen tcp 192.168.1.105:8809: bind: cannot assign requested address"
```

**Comment by jandubois** [comment]
Is the address already in use? Does it work if you specify a different address?

And did you use this address before for a container you have since then deleted? I've seen that ports have not been released properly when the container was destroyed.

**Comment by AkihiroSuda** [comment]
> 192.168.1.105 is my local private ip

Is this VMnet IP?
Possibly a typo of 192.168.105.XXX?
https://github.com/lima-vm/lima/blob/v0.12.0/docs/network.md

**Comment by Taymindis** [comment]
Hi, I've 
- tested `podman` on Mac(MyLocalMachine-NOTVM) and no issue
- tested `containerd` on linux(VM) and no issue
- tested `docker` on Mac(MyLocalMachine-NOTVM) no issue

The only issue is tested `lima ` on (MyLocalMachine-NOTVM) has issue, created new vip also not working

There is no typo

```bash

sh-3.2$ ifconfig en0 | grep 192
	inet 192.168.1.104 netmask 0xffffff00 broadcast 192.168.1.255
sh-3.2$ lima nerdctl run -d --name nginx -p 192.168.1.104:8809:80 nginx:alpine
FATA[0000] failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: time="2022-10-01T02:11:19Z" level=fatal msg="listen tcp 192.168.1.104:8809: bind: cannot assign requested address"
Failed to write to log, write /home/tadika.linux/.local/share/nerdctl/1935db59/containers/default/fc408b59e1b68118cd2ce871f85bac3002954beb7b54bd7b46fb889d9927507b/oci-hook.createRuntime.log: file already closed: unknown
sh-3.2$ sw_vers
ProductName:	macOS
ProductVersion:	12.6
BuildVersion:	21G115


```

**Comment by AkihiroSuda** [comment, marked as answer]
`192.168.1.104` seems the IP of the host, not the guest, so this is an expected behavior.

Maybe you want to try enabling VMnet with the `bridged` mode so that the VM will get a "real" IP address in the 192.168.1.0/24 network: https://github.com/lima-vm/lima/blob/v0.12.0/docs/network.md

**Reply by Taymindis** [reply]
hi @AkihiroSuda , 

It works now! but it only get one and only one IP address in 192.168.1.0/24 segment. How could I get more than 1 ip? 

For eg when enabling bridge mode.

```bash
lima nerdctl run -d --name nginx -p 192.168.1.108:8809:80 nginx:alpine <<< WORKS

lima nerdctl run -d --name nginx -p 192.168.1.109:8809:80 nginx:alpine <<< NOT WORKING


```

**Reply by jandubois** [reply]
You get one IP per MAC address, so if you want multiple IPs, you need to create multiple bridged interfaces. There will be a relatively low limit on the number of interfaces VMNET lets you create though, so don't attempt to use like a 100 different addresses.

Why are you not using different ports on the same address though?

**Reply by Taymindis** [reply]
hi @jandubois 

We actually want to act as multi vm environment to have communication with each other vm. But due to limitation of resource, so I plan to do single VM and multiple IP for those cluster apps testing(Cassandra, kafka and etc ... )


Indeed, for production env, I will need just `one` bridge interface will do. :)

`you need to create multiple bridged interfaces` 

 - How could i create multiple bridge interface? Basically i just need 2 or 3 interface only, did you mean start multiple lima instance?
 - How would I know which IP does it assigned to? It doesn't show in my network interface, i am just blindly guess.

**Reply by jandubois** [reply]
> We actually want to act as multi vm environment to have communication with each other vm.

If you only want to communicate between VMs and not be addressable from the local network, then you should use a `shared` network, not a `bridged` one.

> * How could i create multiple bridge interface? Basically i just need 2 or 3 interface only, did you mean start multiple lima instance?

You can add 3 bridged interfaces to a single instance in `lima.yaml`:

```yaml
networks:
- lima: bridged
- lima: bridged
- lima: bridged
```

> * How would I know which IP does it assigned to? It doesn't show in my network interface, i am just blindly guess.

Run `lima ip a`; the interfaces will be called `lima0`, `lima1`, and `lima2` by default if you don't assign names explicitly.

```console
$ for i in 0 1 2; do lima ip a show lima$i | grep 'inet ' | awk '{print $2}'; done
192.168.1.174/16
192.168.1.173/16
192.168.1.172/16
```

**Reply by Taymindis** [reply]
@jandubois 

Thanks for the full details, I have tried but it show ipv6 address.

```bash

3: lima0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:55:55:36:54:fb brd ff:ff:ff:ff:ff:ff
    inet6 fe80::5055:55ff:fe36:54fb/64 scope link
       valid_lft forever preferred_lft forever
4: lima1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:55:55:56:7b:22 brd ff:ff:ff:ff:ff:ff
    altname enp0s3
    inet6 fe80::5055:55ff:fe56:7b22/64 scope link
       valid_lft forever preferred_lft forever
5: lima2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:55:55:3a:e4:59 brd ff:ff:ff:ff:ff:ff
    altname enp0s4
    inet6 fe80::5055:55ff:fe3a:e459/64 scope link
       valid_lft forever preferred_lft forever
```

**Reply by jandubois** [reply]
I guess it means your local DHCP server is not providing it with an IPv4 address; not sure why that happens. Could be a configuration issue on your networking equipment.

Or (more likely) you are bridging through the wrong host interface. Check `~/.lima/_config/networks.yaml`; it will define the `bridged` network on `en0` (typically the Ethernet interface):

```yaml
networks:
  shared:
    mode: shared
    gateway: 192.168.105.1
    dhcpEnd: 192.168.105.254
    netmask: 255.255.255.0
  bridged:
    mode: bridged
    interface: en0
```

If that interface isn't connected to the local network, then you won't be able to bridge through it either. Try `en1`, which is usually the name of the Wifi interface. Or try this command to find the highest priority interface that does have network connectivity (assumes you have the `jq` tool installed):

```console
$ system_profiler SPNetworkDataType -json | jq -r 'first(.SPNetworkDataType[] | select(.ip_address) | .interface)'
en0
```

Or use "shared" networks, so you don't have to care about all this (but won't be able to connect to your VM from outside the host).

**Reply by Taymindis** [reply]
This is how it looks like. nope we can't use shared as in future need to cross network communication?

```bash

sh-3.2$ system_profiler SPNetworkDataType -json | jq -r 'first(.SPNetworkDataType[] | select(.ip_address) | .interface)'
en0
sh-3.2$ cat ~/.lima/_config/networks.yaml
# Path to socket_vmnet executable. Because socket_vmnet is invoked via sudo it should be
# installed where only root can modify/replace it. This means also none of the
# parent directories should be writable by the user.
#
# The varRun directory also must not be writable by the user because it will
# include the socket_vmnet pid file. Those will be terminated via sudo, so replacing
# the pid file would allow killing of arbitrary privileged processes. varRun
# however MUST be writable by the daemon user.
#
# None of the paths segments may be symlinks, why it has to be /private/var
# instead of /var etc.
paths:
# socketVMNet requires Lima >= 0.12 .
# socketVMNet has precedence over vdeVMNet.
  socketVMNet: /opt/socket_vmnet/bin/socket_vmnet
# vdeSwitch and vdeVMNet are DEPRECATED.
  vdeSwitch: /opt/vde/bin/vde_switch
  vdeVMNet: /opt/vde/bin/vde_vmnet
  varRun: /private/var/run/lima
  sudoers: /private/etc/sudoers.d/lima

group: everyone

networks:
  shared:
    mode: shared
    gateway: 192.168.105.1
    dhcpEnd: 192.168.105.254
    netmask: 255.255.255.0
  bridged:
    mode: bridged
    interface: en0
    # bridged mode doesn't have a gateway; dhcp is managed by outside network
  host:
    mode: host
    gateway: 192.168.106.1
    dhcpEnd: 192.168.106.254
    netmask: 255.255.255.0
sh-3.2$

```

**Reply by Taymindis** [reply]
@jandubois 

I think i got you, End up i create mulitple `- lima: shared` to using shared ip for inter communication and it works. Thanks!
